corporate/views/remote_billing_page.py

```diff
@@ -332,14 +349,189 @@ def remote_billing_legacy_server_login(
+    # do a hacky approach to change it into the self hosting management subdomain.
+    new_hostname = f"{settings.SELF_HOSTING_MANAGEMENT_SUBDOMAIN}.{settings.EXTERNAL_HOST}"
+    split_url = urlsplit(url)
+    modified_url = split_url._replace(netloc=new_hostname)
+    final_url = urlunsplit(modified_url)
+
+    context = {
+        "remote_server_hostname": remote_server.hostname,
+        "remote_server_uuid": str(remote_server.uuid),
+        "confirmation_url": final_url,
+    }
+    send_email(
+        "zerver/emails/remote_billing_legacy_server_confirm_login",
+        to_emails=[email],
+        from_address=FromAddress.tokenized_no_reply_address(),
+        context=context,
+    )
+
+    return render(
+        request,
+        "corporate/remote_billing_legacy_server_confirm_login_sent.html",
+        context={"email": email},
+    )
+
+
+@self_hosting_management_endpoint
+@typed_endpoint
+def remote_billing_legacy_server_from_login_confirmation_link(
+    request: HttpRequest,
+    *,
+    confirmation_key: PathOnly[str],
+    full_name: Optional[str] = None,
+    tos_consent: Literal[None, "true"] = None,
+) -> HttpResponse:
+    """
+    The user comes here via the confirmation link they received via email.
+    """
+
+    try:
+        prereg_object = get_object_from_key(
+            confirmation_key,
+            [Confirmation.REMOTE_SERVER_BILLING_LEGACY_LOGIN],
+            # These links are re-usable.
+            mark_object_used=False,
+        )
+    except ConfirmationKeyError as exception:
+        return render_confirmation_key_error(request, exception)
+    assert isinstance(prereg_object, PreregistrationRemoteServerBillingUser)
+    remote_server = prereg_object.remote_server
+    remote_server_uuid = str(remote_server.uuid)
+
+    # If this user (identified by email) already did this flow, meaning the have a RemoteServerBillingUser,
+    # then we don't re-do the ToS consent  again.
+    tos_consent_needed = not RemoteServerBillingUser.objects.filter(
+        remote_server=remote_server, email=prereg_object.email
+    ).exists()
+
+    if request.method == "GET":
+        context = {
+            "remote_server_uuid": remote_server_uuid,
+            "host": remote_server.hostname,
+            "user_email": prereg_object.email,
+            "tos_consent_needed": tos_consent_needed,
+            "action_url": reverse(
+                remote_billing_legacy_server_from_login_confirmation_link,
+                args=(confirmation_key,),
+            ),
+            "legacy_server_confirmation_flow": True,
+        }
+        return render(
+            request,
+            # TODO: We're re-using the template, so it should be renamed
+            # to a more general name.
+            "corporate/remote_realm_billing_finalize_login_confirmation.html",
+            context=context,
+        )
+
+    if request.method != "POST":
+        return HttpResponseNotAllowed(["GET", "POST"])
```

Comment:
Hmm yeah would be nicer, but here the idea was to ensure in this specific place that POST is being used before proceeding - because we're starting to process data submitted via the form. If we just do `GET or POST` at the start, then we'll accept these params from `GET` too.  Doesn't seem explicitly dangerous in this case, but feels like it must be a violation of some good practices.

Reply:
I see, OK. But the `GET` code path already returns, so we do know that.

Moved this upwards; `remote_billing_legacy_server_from_login_confirmation_link` now uses the exact same pattern as `remote_realm_billing_finalize_login` for GET/POST checking, with a check at the top and an assert after the `GET` code path may have exited.